hal_core/src/lib.rs

```diff
@@ -7,27 +7,53 @@ pub enum Error {}
 
 pub type TimerCallbackFn = fn();
 
-pub type Range = (usize, usize);
-// /// A range similar to core::ops::Range but that is copyable.
-// /// The range is half-open, inclusive below, exclusive above, ie. [start; end[
-// #[derive(Debug, Copy, Clone, PartialEq)]
-// pub struct Range<T: Copy> {
-//     pub start: T,
-//     pub end: T,
-// }
-//
-// impl<T: Copy + core::cmp::PartialOrd + core::cmp::PartialEq + core::ops::Sub<Output = T>>
-//     Range<T>
-// {
-//     pub fn new(start: T, end: T) -> Self {
-//         Self { start, end }
-//     }
-//
-//     pub fn contains(&self, val: T) -> bool {
-//         self.start <= val && val < self.end
-//     }
-//
-//     pub fn size(&self) -> T {
-//         self.end - self.start
-//     }
-// }
+/// A range similar to core::ops::Range but that is copyable.
+/// The range is half-open, inclusive below, exclusive above, ie. [start; end[
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct AddressRange {
+    pub start: usize,
+    pub end: usize,
+}
+
+impl AddressRange {
+    pub fn new(start: usize, end: usize) -> Self {
+        // assert!(range.start % page_size == 0);
+        // assert_eq!(range.end, mm::align_up(range.end, page_size));
+
+        assert!(start < end);
+
+        Self {
+            start,
+            end,
+        }
+    }
+
+    pub fn with_size(start: usize, size: usize) -> Self {
+        Self::new(start, start + size)
+    }
+
+    pub fn round_up_to_page(self, page_size: usize) -> Self {
+        Self {
+            start: self.start,
+            end: mm::align_up(self.end, page_size),
+        }
+    }
+
+    pub fn iter_pages(self, page_size: usize) -> impl Iterator<Item = usize> {
```

Comment:
I wonder if we should just provide `page_size` to new. Are we using `AddressRange` for something else than pages?

Reply:
We don't use it for anything else than pages, I've checked